presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordSink.java

```diff
@@ -92,29 +102,164 @@
-        connectorSession = handle.getConnectorSession();
-        hasDateTimeTypes = handle.getColumnTypes().stream().map(this::containsDateTime).collect(toList());
+        basePath = target;
+        fieldCount = columnNames.size();
+        dataFieldsCount = dataColumnNames.size();
 
-        Iterable<String> hiveTypeNames = transform(transform(handle.getColumnTypes(), HiveType::toHiveType), HiveType::getHiveTypeName);
+        this.conf = conf;
+        this.isPartitioned = isPartitioned;
 
-        Properties properties = new Properties();
-        properties.setProperty(META_TABLE_COLUMNS, Joiner.on(',').join(handle.getColumnNames()));
+        sampleWeightField = columnNames.indexOf(SAMPLE_WEIGHT_COLUMN_NAME);
+        this.columnTypes = columnTypes;
+        this.connectorSession = connectorSession;
+        hasDateTimeTypes = columnTypes.stream().map(this::containsDateTime).collect(toList());
+
+        Iterable<String> hiveTypeNames = transform(transform(dataColumnTypes, HiveType::toHiveType), HiveType::getHiveTypeName);
+
+        properties = new Properties();
+        properties.setProperty(META_TABLE_COLUMNS, Joiner.on(',').join(dataColumnNames));
         properties.setProperty(META_TABLE_COLUMN_TYPES, Joiner.on(':').join(hiveTypeNames));
 
-        serializer = initializeSerializer(conf, properties, handle.getHiveStorageFormat().getSerDe());
-        recordWriter = createRecordWriter(target, conf, properties, handle.getHiveStorageFormat().getOutputFormat());
+        if (serdeParameters != null) {
+            for (String key : serdeParameters.keySet()) {
+                properties.setProperty(key, serdeParameters.get(key));
+            }
+        }
+
+        try {
+            serializer = (Serializer) lookupDeserializer(serdeLib);
+            Class<?> clazz = Class.forName(outputFormat);
+            outputFormatClass = clazz.asSubclass(HiveOutputFormat.class);
+            serializer.initialize(conf, properties);
+        }
+        catch (ClassNotFoundException | ClassCastException | SerDeException e) {
+            throw Throwables.propagate(e);
+        }
 
-        tableInspector = getStandardStructObjectInspector(handle.getColumnNames(), getJavaObjectInspectors(columnTypes));
+        // many nodes can be writing to same partition so create node specific filename
+        filePrefix = (filePrefix.length() > 0) ? filePrefix + "_" : filePrefix;
+        fileName = filePrefix + randomUUID().toString();
+        filesWritten = new HashMap<String, List<String>>();
+        if (isPartitioned) {
+            recordWriters = CacheBuilder.newBuilder()
+                    .build(
+                            new CacheLoader<String, RecordWriter>() {
+                                @Override
+                                public RecordWriter load(String path)
+                                {
+                                    return getRecordWriter(path);
```

Comment:
Here should use 
try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
    return getRecordWriter(path);
}, 
or in concurrent insert it will cause weird problems.